### ISSUE TITLE
Uncaught Error: expected Array [  B {    __uniqueid: 1,    area: 1,    disjointCache: Array [...

### ISSUE THREAD

**Issue by erelsgl** [tswaters/atom-mocha-runner#9]
[Enter steps to reproduce below:]
1. I created a trivial test file under my test/folder:

```
describe('trivial test', function() {
    it('does nothing', function() {
      console.log("hello mocha")
  })
})
```
1. I clicked Ctrl+Alt+O

**Atom Version**: 1.9.8
**System**: Ubuntu 16.04.1
**Thrown From**: [mocha-runner](https://github.com/tswaters/atom-mocha-runner) package, v1.1.10
### Stack Trace

Uncaught Error: expected Array [
  B {
    __uniqueid: 1,
    area: 1,
    disjointCache: Array [ false ],
    factory: ie {
      coordinateSequenceFactory: qt {},
      precisionModel: ee { modelType: ne { name: 'FLOATING' }, scale: null },
      SRID: 0
    },
    maxx: null,
    maxy: null,
    minx: null,
    miny: null
  }
] to have property length of 2 (got 1)

```
At /home/erelsgl/.atom/packages/mocha-runner/lib/runner.js:47

AssertionError: expected Array [
  B {
    __uniqueid: 1,
    area: 1,
    disjointCache: Array [ false ],
    factory: ie {
      coordinateSequenceFactory: qt {},
      precisionModel: ee { modelType: ne { name: 'FLOATING' }, scale: null },
      SRID: 0
    },
    maxx: null,
    maxy: null,
    minx: null,
    miny: null
  }
] to have property length of 2 (got 1)
    at Assertion.fail (/home/erelsgl/git/computational-geometry/node_modules/should/cjs/should.js:326:17)
    at Assertion.Object.defineProperty.value (/home/erelsgl/git/computational-geometry/node_modules/should/cjs/should.js:398:19)
    at /home/erelsgl/git/computational-geometry/test/MaxDisjointSetAsync-test.js:66:23
    at /home/erelsgl/git/computational-geometry/node_modules/async/dist/async.js:2109:20
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```
### Commands

```
     -0:01.1.0 mocha-runner:run-specs (atom-text-editor.editor.is-focused)
```
### Config

``` json
{
  "core": {
    "audioBeep": false,
    "disabledPackages": [
      "mocha-test-runner",
      "atom-mocha",
      "mocha-ui"
    ],
    "excludeVcsIgnoredPaths": false,
    "projectHome": "/home/erelsgl/git",
    "themes": [
      "atom-light-ui",
      "atom-light-syntax"
    ]
  }
}
```
### Installed Packages

``` coffee
# User
atom-mocha, v0.0.8 (inactive)
build, v0.65.0 (active)
busy, v0.6.0 (active)
mocha-runner, v1.1.10 (active)
mocha-test-runner, v0.5.1 (inactive)
mocha-ui, v0.1.8 (inactive)
node-debug, v0.1.7 (inactive)
node-debugger, v1.7.0 (active)
atom-dark-syntax, v0.27.0 (inactive)
atom-dark-ui, v0.51.0 (inactive)
atom-light-syntax, v0.28.0 (active)
atom-light-ui, v0.43.0 (active)
base16-tomorrow-dark-theme, v1.1.0 (inactive)
base16-tomorrow-light-theme, v1.1.1 (inactive)
one-dark-ui, v1.3.2 (inactive)
one-light-ui, v1.3.2 (inactive)
one-dark-syntax, v1.2.0 (inactive)
one-light-syntax, v1.2.0 (inactive)
solarized-dark-syntax, v1.0.2 (inactive)
solarized-light-syntax, v1.0.2 (inactive)
about, v1.5.2 (active)
archive-view, v0.61.1 (active)
autocomplete-atom-api, v0.10.0 (active)
autocomplete-css, v0.11.1 (active)
autocomplete-html, v0.7.2 (active)
autocomplete-plus, v2.31.0 (active)
autocomplete-snippets, v1.11.0 (active)
autoflow, v0.27.0 (inactive)
autosave, v0.23.1 (active)
background-tips, v0.26.0 (active)
bookmarks, v0.41.0 (active)
bracket-matcher, v0.82.1 (active)
command-palette, v0.38.0 (inactive)
deprecation-cop, v0.54.1 (active)
dev-live-reload, v0.47.0 (active)
encoding-selector, v0.22.0 (active)
exception-reporting, v0.39.0 (active)
fuzzy-finder, v1.3.0 (active)
git-diff, v1.1.0 (active)
find-and-replace, v0.198.0 (inactive)
go-to-line, v0.31.0 (inactive)
grammar-selector, v0.48.1 (active)
image-view, v0.58.0 (active)
incompatible-packages, v0.26.1 (active)
keybinding-resolver, v0.35.0 (active)
line-ending-selector, v0.5.0 (active)
link, v0.31.1 (inactive)
markdown-preview, v0.158.0 (active)
metrics, v0.53.1 (active)
notifications, v0.64.1 (active)
open-on-github, v1.2.0 (inactive)
package-generator, v1.0.0 (inactive)
settings-view, v0.238.2 (active)
snippets, v1.0.2 (active)
spell-check, v0.67.1 (active)
status-bar, v1.4.1 (active)
styleguide, v0.46.0 (active)
symbols-view, v0.113.0 (inactive)
tabs, v0.98.1 (active)
timecop, v0.33.1 (active)
tree-view, v0.208.0 (active)
update-package-dependencies, v0.10.0 (active)
welcome, v0.34.0 (active)
whitespace, v0.32.2 (active)
wrap-guide, v0.38.1 (active)
language-c, v0.52.1 (active)
language-clojure, v0.21.0 (active)
language-coffee-script, v0.47.0 (active)
language-csharp, v0.12.1 (active)
language-css, v0.36.2 (active)
language-gfm, v0.86.0 (active)
language-git, v0.13.0 (active)
language-go, v0.42.0 (active)
language-html, v0.44.1 (active)
language-hyperlink, v0.16.0 (active)
language-java, v0.22.0 (active)
language-javascript, v0.119.0 (active)
language-json, v0.18.0 (active)
language-less, v0.29.3 (active)
language-make, v0.22.2 (active)
language-mustache, v0.13.0 (active)
language-objective-c, v0.15.1 (active)
language-perl, v0.35.0 (active)
language-php, v0.37.0 (active)
language-property-list, v0.8.0 (active)
language-python, v0.45.0 (active)
language-ruby, v0.68.5 (active)
language-ruby-on-rails, v0.25.0 (active)
language-sass, v0.52.0 (active)
language-shellscript, v0.22.3 (active)
language-source, v0.9.0 (active)
language-sql, v0.21.1 (active)
language-text, v0.7.1 (active)
language-todo, v0.28.0 (active)
language-toml, v0.18.0 (active)
language-xml, v0.34.8 (active)
language-yaml, v0.26.0 (active)

# Dev
No dev packages
```

**Comment by tswaters** [comment]
Are the contents of `MaxDisjointSetAsync-test.js` just that dummy does nothing test?

It looks like the runner picked up additional tests -- still it shouldn't crash.

What are the contents of mocha.opts?  Is this repo on github so I can clone and reproduce?

I'll look into this further in the coming days...

**Comment by erelsgl** [comment]
I wanted to create a MWE (minimal-working-example) for you to make it
easier for you to find the error. So I created a folder "test2" with a
single file "trivial.js" with the trivial test I described above.

I also changed the "Test directory" in the Settings from the default "test"
to "test2".  Then, I clicked Ctrl+Alt+O while standing inside the file
test2/trivial.js, hoping that only this test will be executed.

Apparently, this did not work: all test from the "test/" folder were
executed.

Is there a way I call tell mocha-runner to run only a single test-file?
This will make it much easier for me to create a MWE.

_--- Erel Segal-Halevi   (http://erelsgl.github.io
http://erelsgl.github.io) אראל סגל הלוי_

**Comment by tswaters** [comment]
That sounds like another bug... sigh.

You can tell mocha to run just one file by including a mocha.opts file pointing at the file

./test/MWE.js
./test/mocha.opts

In mocha.opts, one line:

test/MWE.js

FWIW, it will probably work there.... it appears there is an assertion error in your tests that isn't handled properly.  If I were to guess, I'd say it relates to should but I'm not at my home workstation and can't verify. 

I'll look into this later, probably after I get back home from work, ~8 hours or so.

**Comment by tswaters** [comment]
Ok, I figured it out.  The problem is when there is an async test that throws an exception.  

To ease my development, I added some code to bubble errors from the child process back to the main atom process so I could see errors.  Reason being the debugger / dev tools can't be attached to a child process and you're basically flying blind.... as best as I can remember the only reason it was there was for my own utility during development [[1](https://github.com/tswaters/atom-mocha-runner/blob/v1.1.10/lib/runner.js#L42-L48)], [[2](https://github.com/tswaters/atom-mocha-runner/blob/v1.1.10/lib/worker/index.js#L14-L29)]

So, mocha [adds it's own uncaughtException handler](https://github.com/mochajs/mocha/blob/v2.5.3/lib/runner.js#L803) which basically handles the case where an async test throws an error.  Mine was attached first, so ran first and bubbled up the error to the atom process and showed the error.  I'm going to remove my handler - any exceptions thrown in async tests should be handled by the mocha reporter.

This is fixed in [v1.1.11](https://github.com/tswaters/atom-mocha-runner/releases/tag/v1.1.11)
